Forward AD formulas batch 1

albanD opened the pull request:
Note that this PR implements formulas only for ops that are supported by OpInfo.
Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58094 Forward AD formulas batch 3
* #57863 Forward AD formulas batch 2
* **#57768 Forward AD formulas batch 1**

Differential Revision: [D28387766](https://our.internmc.facebook.com/intern/diff/D28387766)

albanD committed:
Forward AD formulas batch 1

[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.




[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.




[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.




[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.




[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.




[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.




[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.




[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.




[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387766).

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.
Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387766](https://our.internmc.facebook.com/intern/diff/D28387766)

[ghstack-poisoned]

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.
Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387766](https://our.internmc.facebook.com/intern/diff/D28387766)

[ghstack-poisoned]

zou3519 commented on tools/autograd/load_derivatives.py:
```diff
@@ -169,10 +169,11 @@ def find_required_inputs(formula: str, postfix: str) -> Tuple[str, ...]:
                                    "defines the gradient formula for its argument which is required.")
             # This transformation is based on the observation that for element-wise functions, the Jacobian
-            # matrix is diagonal and thus doing J * v or v * J gives the same result.
+            # matrix is diagonal and thus doing J * v or v * J.conj() or (v.conj() J).conj() gives the same result.
```
I agree that `v * J.conj()` and `(v.conj() J).conj()` are the same, but are those really the same as `J * v` ?

soulitzer replied:
I think it meant to be `J.conj() * v`

albanD replied:
No the idea is really to say that we want to do `J * v` and every transposition is replaced by a hermitian transpose.
Since J is diagonal, it doesn't move but v and the result are hermitian transposed, hence the `(v^H J)^H`. Ignoring the tranpose again leaves just the conjugates.

zou3519 commented on tools/autograd/load_derivatives.py:
```diff
@@ -169,10 +169,11 @@ def find_required_inputs(formula: str, postfix: str) -> Tuple[str, ...]:
+            # 1) all occurrences of "grad" with "foo_t.conj()", where foo is the name of the unique differentiable input.
             # 2) all usage of an original input "foo" with its primal value "foo_p".
+            # 3) conjugate the final result
             # For example, for abs, the backward formula is:
             #   grad * self.sgn()
             # And this function generates a forward formula that is:
```
Should the example be updated?

albanD replied:
Yes that needs to be updated for complex, good point!

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.
Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387766](https://our.internmc.facebook.com/intern/diff/D28387766)

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5597,6 +5615,7 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
                    decorators=(precisionOverride({torch.float16: 1e-2,
                                                   torch.bfloat16: 1e-2}),),
                    safe_casts_outputs=True,
+                   supports_forward_ad=True,
```
Are the following three operators tested in this PR?
```
alias
add.Scalar
_sparse_addmm
```
I know the formulas for alias and add.Scalar are really simple, but still would be good to have test coverage

soulitzer replied:
I don't see _sparse_addmm added in this PR, it might be the github UI hiding the lines :P
For alias/add.Scalar I agree that it would be good to have those covered, though I'm guessing there aren't any OpInfos yet for that yet?

zou3519 replied:
Good point, git blame confused me -- _sparse_addmm is not actually there :)

alias/add.Scalar are only accessible from the C++ API, so they can't really be tested on the Python side...

albanD replied:
Yes, they cannot be accessed from python :/ 
The two formulas are trivial though so I think its ok to keep them no?

zou3519 replied:
I'm fine with merging this without tests for alias and add because those are very simple

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -220,20 +226,24 @@
 
 - name: alias(Tensor(a) self) -> Tensor(a)
   self: grad
+  result: auto_element_wise
```
Is `self_t` more efficient? Otherwise the codegen will name this do `self_t.conj().conj()`, right?

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -190,26 +191,31 @@
   self: handle_r_to_c(self.scalar_type(), grad)
   tensor1: handle_r_to_c(tensor1.scalar_type(), grad * (value / tensor2).conj())
   tensor2: handle_r_to_c(tensor2.scalar_type(), -grad * (value * tensor1 / (tensor2 * tensor2)).conj())
+  result: self_t + maybe_multiply(tensor1_t / tensor2_p, value) - maybe_multiply(tensor2_t * (tensor1_p / tensor2_p) / tensor2_p, value)
```
(no action required) Could you actually "auto-elementwise" this and other pointwise operations? It looks like the formula (for the real case at least) is just the backward formula for self + backward formula for tensor 1 + backward formula for tensor2 while replacing all the grads with the correct tangents.

albanD replied:
Yes we could do it here. Will add it if it shows up again.

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -190,26 +191,31 @@
   self: maybe_multiply(grad, beta.conj())
   mat1: mm_mat1_backward(grad, mat2, mat1.sizes(), mat1.strides(), alpha)
   mat2: mm_mat2_backward(grad, mat1, mat2.sizes(), mat2.strides(), alpha)
+  result: maybe_multiply(self_t, beta) + maybe_multiply(mat1_t.mm(mat2_p), alpha) + maybe_multiply(mat1_p.mm(mat2_t), alpha)
```
It's interesting to note that this is just `maybe_multiply(self_t, beta)` added with `maybe_multiply( formula_for_mm , alpha )`. Are there any chances we would want to dedup code between this and the mm formula in the future?

albanD replied:
The main problem with such formula is that they are not element-wise. So adding the formulas won't work.
And they are affine (not linear) and so we would need to provide some arguments to a smarter auto_affine to handle this. Which feels like a dangerous step to take.

zou3519 replied:
Yeah, I agree the design for this would be tricky.

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -220,20 +226,24 @@
 
 - name: angle(Tensor self) -> Tensor
   self: angle_backward(grad, self)
+  result: handle_r_to_c(result.scalar_type(), angle_backward(self_t, self_p))
```
Why do we need  `handle_r_to_c` here?

zou3519 replied:
(resolved offline)

zou3519 reviewed (commented):
the formulas lgtm from a real numbers perspective, but I am not sure how to derive them for complex numbers

zou3519 reviewed (approved):
Offline Alban walked me through a derivation of complex forward-mode AD derivative for torch.sin and torch.conj and those helped me understand enough to derive the complex formulas as well.

NB: we should update the one example above that wasn't updated for this PR, but other than that things lgtm

albanD committed:
Update on "Forward AD formulas batch 1"

Note that this PR implements formulas only for ops that are supported by OpInfo.
Slow gradcheck also passes for this PR and can be found here: https://github.com/pytorch/pytorch/pull/57976


Differential Revision: [D28387766](https://our.internmc.facebook.com/intern/diff/D28387766)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28387766).
